services/database/user.ts

```diff
@@ -35,4 +41,177 @@ async function getUsers() {
     return users;
 }
 
-export { getUser, getUsers };
+/* 
+TODO (6/30/21): test passing userId in newTeacherData instead of through connect key
+*/
+
+async function updateUser(user: UserInput): Promise<User> {
+    /*
+    Flow:
+    - get role and role data from UserInput
+    - create/update the role record
+    - update the user record
+    - return the user, including the role records
+    */
+    const roleData = user.role_data;
+    switch (user.role) {
+        case "PARENT": {
+            upsertParent(roleData, user.id);
+            break;
+        }
+        case "PROGRAM_ADMIN": {
+            upsertProgramAdmin(roleData, user.id);
+            break;
+        }
+        case "VOLUNTEER": {
+            upsertVolunteer(roleData, user.id);
+            break;
+        }
+    }
+
+    const updatedUser = await prisma.user.update({
+        data: {
+            first_name: user.first_name,
+            last_name: user.last_name,
+        },
+        where: { id: parseInt(user.id) },
+        include: {
+            teachers: true,
+            parents: true,
+            program_admins: true,
+            volunteers: true,
+        },
+    });
+    return updatedUser;
+}
+
+/**
+ * Creates a new teacher corresponding to newTeacherData and
+ * returns the newly created teacher
+ * @param newTeacherData - data corresponding to the new teacher user
+ * @returns - the newly created teacher user
+ */
+async function upsertTeacher(
+    teacherData: TeacherInput,
+    userId: string,
+): Promise<Teacher> {
+    const id: number = parseInt(userId);
+    const upsertTeacherData = {
+        ...teacherData,
+        id: id,
+    };
+    const teacher = await prisma.teacher.upsert({
+        create: upsertTeacherData,
+        update: upsertTeacherData,
+        where: { id: id },
+    });
+    return teacher;
+}
+
+/**
+ * Creates a new volunteer corresponding to newVolunteerData
+ * and returns the newly created volunteer
+ * @param newVolunteerData - data corresponding to the new volunteer volunteer user
+ * @returns - the newly created volunteer user
+ */
+async function upsertVolunteer(
+    volunteerData: VolunteerInput,
+    userId: string,
+): Promise<Volunteer> {
+    const id: number = parseInt(userId);
+    const upsertVolunteerData = {
+        ...volunteerData,
+        id: id,
+    };
+    const volunteer = await prisma.volunteer.upsert({
+        create: upsertVolunteerData,
+        update: upsertVolunteerData,
+        where: { id: id },
+    });
+    return volunteer;
+}
+
+/**
+ * TODO (July 1):
```

Comment:
@sohak22 I think I (may have?) figured out how to use the upsert function properly and why we were getting errors with the data key in the create/update functions - I think it's because we were trying to pass in `id` as a string when it should be a number.

I started updating some of the functions (`upsertTeacher`, `upsertVolunteer`), if you have time today or tomorrow you can maybe give `upsertParent` and `upsertProgramAdmin` a try. Let me know if you have questions (about how to implement or why I made the changes that I did)!! also idek if I did this correctly so fingers crossed

Reply:
Yeah, you guys are right that it expects a number so you need to have `parseInt(id)`. Will take a look at this PR today!